### PR TITLE
Fix/stuck on one bet

### PR DESCRIPTION
**Pull request by annasambrook**
Fix for the issue of sampling getting stuck on one unprofitable bet.

### CODE REVIEW

File: packages/valory/skills/decision_maker_abci/behaviours/bet_placement.py
Comment (Adamantios) [inline packages/valory/skills/decision_maker_abci/behaviours/bet_placement.py:1]: Can you explain these changes?

File: pyproject.toml [hunk 1]
```diff
@@ -17,7 +17,7 @@ include = "packages"
 [tool.poetry.dependencies]
 python = "<3.12,>=3.8"
 open-autonomy = "==0.14.14.post1"
-requests = "==2.28.1"
+requests = "<2.31.2,>=2.28.1"
```
Comment (Adamantios) [inline pyproject.toml:20]: Did you face some issues with the version?

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py [hunk 2]
```diff
@@ -95,6 +125,13 @@ def _sample(self) -> Optional[int]:
         """Sample a bet, mark it as processed, and return its index."""
         available_bets = list(filter(self.processable_bet, self.bets))
 
+        # store the bet selection stats for each processable bet at the time of sampling
+        for bet in available_bets:
+            self.shared_state.bet_selection_stats[bet.id] = {
+                "scaled_liquidity_measure": bet.scaledLiquidityMeasure,
+                "outcome_token_amounts": bet.outcomeTokenAmounts,
+            }
```
Comment (Adamantios) [inline packages/valory/skills/decision_maker_abci/behaviours/sampling.py:133]: Please convert the nested dict to a dataclass to avoid human error.

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py [hunk 3]
```diff
@@ -118,7 +155,6 @@ def async_act(self) -> Generator:
         """Do the action."""
         with self.context.benchmark_tool.measure(self.behaviour_id).local():
             idx = self._sample()
-            self.store_bets()
```
Comment (Adamantios) [inline packages/valory/skills/decision_maker_abci/behaviours/sampling.py:121]: Why removed?

File: packages/valory/skills/market_manager_abci/bets.py [hunk 4]
```diff
@@ -108,7 +108,16 @@ def __post_init__(self) -> None:
 
     def __lt__(self, other: "Bet") -> bool:
         """Implements less than operator."""
-        return self.scaledLiquidityMeasure < other.scaledLiquidityMeasure
+        outcome_index = self.prediction_response.vote
+
+        if self.n_bets > 0 and outcome_index:
```
Comment (Adamantios) [inline packages/valory/skills/market_manager_abci/bets.py:113]: What if the vote is `0`?

File: packages/valory/skills/market_manager_abci/bets.py [hunk 5]
```diff
@@ -108,7 +108,16 @@ def __post_init__(self) -> None:
...
+            return (
+                self.outcomeTokenAmounts[outcome_index]
+                < other.outcomeTokenAmounts[outcome_index]
+            )
```
Comment (Adamantios) [inline packages/valory/skills/market_manager_abci/bets.py:117]: Why prioritize the markets that have more tokens for the position that we previously voted for?

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py [hunk 6]
```diff
@@ -65,19 +65,49 @@ def processable_bet(self, bet: Bet) -> bool:
...
+                    outcome_liquidity_change = (
+                        bet.outcomeTokenAmounts[outcome_index]
+                        != self.shared_state.bet_selection_stats.get(bet_id, {}).get(
+                            "outcome_token_amounts", [0, 0]
+                        )[outcome_index]
+                    )
+                    return within_ranges and can_rebet and outcome_liquidity_change
+                else:
+                    # market cannot be processed as it doesn't actually have a previous bet
+                    return False
```
Comment (Adamantios) [inline packages/valory/skills/decision_maker_abci/behaviours/sampling.py:110]: This is very hard to read, why did you change the early return pattern?

Comment (Adamantios) [inline packages/valory/skills/decision_maker_abci/behaviours/sampling.py:110]: How can this happen since we have already checked the `n_bets` above?

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py [hunk 7]
```diff
@@ -65,19 +65,49 @@ def processable_bet(self, bet: Bet) -> bool:
...
+                if bet.processed_timestamp > 0:
+                    scaled_liquidity_changed = (
+                        bet.scaledLiquidityMeasure
+                        != self.shared_state.bet_selection_stats.get(bet_id, {}).get(
+                            "scaledLiquidityMeasure", 0
+                        )
+                    )
+                    return within_ranges and can_rebet and scaled_liquidity_changed
+                else:
+                    return within_ranges and can_rebet
```
Comment (Adamantios) [inline packages/valory/skills/decision_maker_abci/behaviours/sampling.py:89]: Why check for `can_rebet` since we are inside the `if not self.should_rebet:` condition?

**Review by Adamantios — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (annasambrook) [reply pyproject.toml:20]: No, I didn't change this myself. I think it got changed when running a command. I can revert this.

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py (hunk 3 above)
Comment (annasambrook) [reply packages/valory/skills/decision_maker_abci/behaviours/sampling.py:121]: This got removed because if the agent crashes or is stopped in between sampling and the bet_placement/blacklisting the n_bets becomes inaccurate and you then face an issue in using bet.prediction_response.vote.

File: packages/valory/skills/market_manager_abci/bets.py (hunk 4 above)
Comment (annasambrook) [reply packages/valory/skills/market_manager_abci/bets.py:113]: ah this needs to be outcome_index is not None

File: packages/valory/skills/market_manager_abci/bets.py (hunk 5 above)
Comment (annasambrook) [reply packages/valory/skills/market_manager_abci/bets.py:117]: This was suggested by Juli.

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py (hunk 2 above)
Comment (annasambrook) [reply packages/valory/skills/decision_maker_abci/behaviours/sampling.py:133]: ok

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py (hunk 7 above)
Comment (annasambrook) [reply packages/valory/skills/decision_maker_abci/behaviours/sampling.py:89]: As Juli and I had understood, can_rebet checks that once a bet has been sampled and either been bet upon or deemed unprofitable , enough time has passed that it can be resampled.

File: packages/valory/skills/decision_maker_abci/behaviours/sampling.py (hunk 6 above)
Comment (annasambrook) [reply packages/valory/skills/decision_maker_abci/behaviours/sampling.py:110]: this was due to a failure when running tomte check-code

Comment (annasambrook) [reply packages/valory/skills/decision_maker_abci/behaviours/sampling.py:110]: this was because of the issue with storing bets in the sampling behaviour before the bet had been placed. this should now be able to be removed as I have moved store_bets

File: packages/valory/skills/decision_maker_abci/behaviours/bet_placement.py
Comment (annasambrook) [reply packages/valory/skills/decision_maker_abci/behaviours/bet_placement.py:1]: we need to store the increase in n_bets only once a bet has been increased to avoid errors if the agent is stopped mid process

**Comment by Adamantios** [comment]
Closed in favor of #341.
